File: atcoder/beginner_contest/151_20200112/d.py
import itertools
import queue
import copy


def calc_dist(S, points, p_s, p_e):
    h, w = len(S), len(S[0])
    distance = {p: -1 for p in points}
    que = queue.Queue()
    que.put((p_s, 0))
    while not que.empty():
        p, d = que.get()
        y, x = p

        if not (0 <= x < w and 0 <= y < h):
            continue
        if S[y][x] == '#':
            continue

        if distance[p] != -1:
            continue

        distance[p] = d
        if p == p_e:
            break

        que.put(((y, x - 1), d + 1))
        que.put(((y, x + 1), d + 1))
        que.put(((y - 1, x), d + 1))
        que.put(((y + 1, x), d + 1))
    
    # print(p_s, p_e, distance[p_e])
    # Svis = copy.deepcopy(S)
    # Svis[p_s[0]][p_s[1]] = 'S'
    # Svis[p_e[0]][p_e[1]] = 'E'
    # for e in Svis:
    #     print(''.join(e))
    return distance[p_e]


def main():
    h, w = map(int, input().split())

    S = list()
    for y in range(h):
        S.append(list(input()))
    points = [(y, x) for y, x in itertools.product(range(h), range(w)) if S[y][x] == '.']

    # print(calc_dist(S, points, (0, 2), (1, 0)))
    distance = list()
    for p_s, p_e in itertools.combinations(points, 2):
        d = calc_dist(S, points, p_s, p_e)
        distance.append(d)

    print(max(distance))


main()

File: atcoder/beginner_contest/151_20200112/d.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <map>
#include <utility>
#include <algorithm>

using namespace std;


int calc_max_dist(vector<string> S, vector<pair<int, int>> points, pair<int, int> p_s) {
    int h = S.size();
    int w = S[0].size();
    map<pair<int, int>, int> distance;
    for (auto p: points) {
        distance[p] = -1;
    }
    queue<pair<pair<int, int>, int>> que;
    que.push(make_pair(p_s, 0));
    while (!que.empty()) {
        auto t = que.front();
        auto p = t.first;
        auto d = t.second;
        que.pop();
        auto y = p.first; auto x = p.second;
        if (0 > x || x >=  w || y < 0 || y >= h) {
            continue;
        }
        if (S[y][x] == '#') {
            continue;
        }
        if (distance[p] != -1) {
            continue;
        }

        distance[p] = d;
        que.push(make_pair(make_pair(y, x - 1), d + 1));
        que.push(make_pair(make_pair(y, x + 1), d + 1));
        que.push(make_pair(make_pair(y - 1, x), d + 1));
        que.push(make_pair(make_pair(y + 1, x), d + 1));
    }
   
    int max_dist = -1;
    for (auto d: distance) {
        max_dist = max(max_dist, d.second);
    }
    return max_dist;
}

int main(int argc, char const *argv[])
{
    int h, w;
    cin >> h >> w;

    vector<string> S(h);
    for (int y = 0; y < h; y++) {
        cin >> S[y];
    }

    vector<pair<int, int>> points;
    for (int y = 0; y < h; y++) {
        for (int x = 0; x < w; x++) {
            if (S[y][x] == '.') {
                points.push_back(make_pair(y, x));
            }
        }
    }
    vector<int> distance;
    for (int i = 0; i < points.size(); i++) {
        auto p_s = points[i];
        int d = calc_max_dist(S, points, p_s);
        distance.push_back(d);
    }
    cout << *max_element(distance.begin(), distance.end()) << endl;
    return 0;
}
